tests/client_api.cpp

```diff
@@ -83,6 +83,54 @@ TEST(ClientAPI, LoginWrongUsername)
         mtx_client->close();
 }
 
+TEST(ClientAPI, EmptyUserAvatar)
+{
+        auto alice = std::make_shared<Client>("localhost");
+
+        alice->login("alice", "secret", [alice](const mtx::responses::Login &res, ErrType err) {
+                ASSERT_FALSE(err);
+
+                auto const alice_id = res.user_id;
+                validate_login(alice_id.toString(), res);
+
+                alice->set_avatar_url("", [](ErrType err) { ASSERT_FALSE(err); });
+
+                alice->download_user_avatar(alice_id,
+                                            [](const mtx::responses::Profile &res, ErrType err) {
+                                                    ASSERT_FALSE(err);
+                                                    ASSERT_TRUE(res.avatar_url.size() == 0);
+                                            });
+
+        });
+
+        alice->close();
+}
+
+TEST(ClientAPI, RealUserAvatar)
+{
+        auto alice = std::make_shared<Client>("localhost");
+
+        alice->login("alice", "secret", [alice](const mtx::responses::Login &res, ErrType err) {
+                ASSERT_FALSE(err);
+
+                auto const alice_id   = res.user_id;
+                auto const avatar_url = "mxc://matrix.org/wefh34uihSDRGhw34";
+
+                validate_login(alice_id.toString(), res);
+
+                alice->set_avatar_url(avatar_url, [](ErrType err) { ASSERT_FALSE(err); });
+
+                alice->download_user_avatar(
```

Comment:
All the api calls are async, so the order of the callbacks here is not guaranteed. You could call `download_user_avatar` inside `set_avatar_url`'s callback.